tracing/tracing-test.sh

```diff
@@ -0,0 +1,247 @@
+
+	[ "$count" -lt "$min_spans" ] && \
+		die "expected >= $min_spans reported spans, got $count"
+
+	info "All $count spans reported"
+
+	rm -f "$logfile"
+}
+
+# Check Jaeger spans for the specified service.
+check_jaeger_status()
+{
+	local service="$1"
+	local min_spans="$2"
+
+	[ -z "$service" ] && die "need jaeger service name"
+	[ -z "$min_spans" ] && die "need minimum trace span count"
+
+	local status
+	local errors=0
+
+	local attempt=0
+	local attempts=3
+
+	while [ "$attempt" -lt "$attempts" ]
+	do
+		status=$(get_jaeger_status "$service")
+
+		#------------------------------
+		# Basic sanity checks
+		[ -z "$status" ] && die "failed to query status via HTTP"
+
+		local span_lines=$(echo "$status"|jq -S '.data[].spans | length')
+		[ -z "$span_lines" ] && die "no span status"
+
+		local span_lines_count=$(echo "$span_lines"|wc -l)
+
+		# Total up all span counts
+		local spans=$(echo "$span_lines"|paste -sd+ -|bc)
+		[ -z "$spans" ] && die "no spans"
+
+		# Ensure total span count is numeric
+		echo "$spans"|grep -q "^[0-9][0-9]*$"
+		[ $? -eq 0 ] || die "invalid span count: '$spans'"
+
+		info "found $spans spans (across $span_lines_count traces)"
+
+		# Validate
+		[ "$spans" -lt "$min_spans" ] && die "expected >= $min_spans spans, got $spans"
+
+		# Look for common errors in span data
+		local errors1=$(echo "$status"|jq -S . 2>/dev/null|grep "invalid parent span")
+		if [ -n "$errors1" ]
+		then
+			errors=$((errors+1))
+			warn "Found invalid parent span errors (attempt $attempt): $errors1"
+			continue
+		else
+			errors=$((errors-1))
+			[ "$errors" -lt 0 ] && errors=0
+		fi
+
+		# Crude but it works
+		local errors2=$(echo "$status"|jq -S . 2>/dev/null|grep "\"warnings\""|grep -E -v "\<null\>")
+		if [ -n "$errors2" ]
+		then
+			errors=$((errors+1))
+			warn "Found warnings (attempt $attempt): $errors2"
+			continue
+		else
+			errors=$((errors-1))
+			[ "$errors" -lt 0 ] && errors=0
+		fi
+
+		attempt=$((attempt+1))
+
+		[ "$errors" -eq 0 ] && break
+	done
+
+	[ "$errors" -gt 0 ] && die "errors still detected after $attempts attempts"
+}
+
+run_test()
+{
+	local min_spans="$1"
+	local service="$2"
+
+	[ -z "$min_spans" ] && die "need minimum span count"
+	[ -z "$service" ] && die "need service name"
+
+	start_jaeger
+
+	create_trace
+
+	check_jaeger_status "$service" "$min_spans"
+
+	check_missing_spans "$service" "$min_spans"
+
+	stop_jaeger
+
+	info "test passed"
+}
+
+main()
+{
+	# FIXME: this will need to be specified per component
+	# (kata-runtime, kata-agent, etc).
+	runtime_min_spans=10
+	shim_min_spans=10
+
+	# Name of Jaeger "service" (aka Kata component) to check trace for.
+	runtime_service="kata-runtime"
+	shim_service="kata-shim"
+
+	[ -n "$1" ] && min_spans="$1"
+	[ "$min_spans" -le 0 ] && die "minimum spans must be positive"
+
+	.ci/configure_tracing_for_kata.sh enable
+	run_test "$runtime_min_spans" "$runtime_service"
+	run_test "$shim_min_spans" "$shim_service"
+	.ci/configure_tracing_for_kata.sh disable
```

Comment:
I was contemplating if we needed the unconfigure of the config in a TRAP function - but, given that if this test dies, the CI will stop running anyway, I suspect we can just leave the file configured for jeager at that point, as the whole test setup will be thrown away and re-created for the next test run...

Reply:
added trap function